cdap-data-fabric/src/main/java/co/cask/cdap/spi/data/table/StructuredTableSpecification.java

```diff
@@ -0,0 +1,237 @@
+  }
+
+  /**
+   * @return the table id of this table specification
+   */
+  public StructuredTableId getTableId() {
+    return tableId;
+  }
+
+  /**
+   * @return the field types of the table
+   */
+  public List<FieldType> getFieldTypes() {
+    return fieldTypes;
+  }
+
+  /**
+   * @return the list of primary keys defined on the table
+   */
+  public List<String> getPrimaryKeys() {
+    return primaryKeys;
+  }
+
+  /**
+   * @return the list of indexes defined on the table
+   */
+  public List<String> getIndexes() {
+    return indexes;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    StructuredTableSpecification that = (StructuredTableSpecification) o;
+    return Objects.equals(tableId, that.tableId) &&
+      Objects.equals(fieldTypes, that.fieldTypes) &&
+      Objects.equals(primaryKeys, that.primaryKeys) &&
+      Objects.equals(indexes, that.indexes);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(tableId, fieldTypes, primaryKeys, indexes);
+  }
+
+  @Override
+  public String toString() {
+    return "StructuredTableSpecification{" +
+      "tableId='" + tableId + '\'' +
+      ", fieldTypes=" + fieldTypes +
+      ", primaryKeys=" + primaryKeys +
+      ", indexes=" + indexes +
+      '}';
+  }
+
+  /**
+   * Builder used to create {@link StructuredTableSpecification}
+   */
+  public static final class Builder {
+    private static final String[] EMPTY_STRING_ARRAY = new String[0];
+
+    private StructuredTableId tableId;
+    private FieldType[] fieldTypes;
+    private String[] primaryKeys;
+    private String[] indexes = EMPTY_STRING_ARRAY;
+
+    /**
+     * Set the table id. A table should have an id.
+     * @param id table id
+     * @return Builder instance
+     */
+    public Builder withId(StructuredTableId id) {
+      this.tableId = id;
+      return this;
+    }
+
+    /**
+     * Set the field types in the table schema. A table should have at least one field.
+     * @param fieldTypes list of field types
+     * @return Builder instance
+     */
+    public Builder withFields(FieldType ...fieldTypes) {
+      this.fieldTypes = fieldTypes;
+      return this;
+    }
+
+    /**
+     * Set the fields that form the primary keys of the table. A table should have at least one primary key.
+     * @param primaryKeys list of field names forming the primary keys
+     * @return Builder instance
+     */
+    public Builder withPrimaryKeys(String ...primaryKeys) {
+      this.primaryKeys = primaryKeys;
+      return this;
+    }
+
+    /**
+     * Set the fields that need to be indexed in the table. A table need not define any indexes.
+     * @param indexes list of field names for the index
+     * @return Builder instance
+     */
+    public Builder withIndexes(String ...indexes) {
+      if (indexes != null) {
```

Comment:
Should this check be part of validate method?

Reply:
See discussion https://github.com/cdapio/cdap/pull/10874#discussion_r242245288